pkg/metrics/metrics_test.go

```diff
@@ -0,0 +1,520 @@
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver4",operation_name="failure2",operation_status="Failure",le="600"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver4",operation_name="failure2",operation_status="Failure",le="+Inf"} 1
+snapshot_controller_operation_total_seconds_sum{driver_name="driver4",operation_name="failure2",operation_status="Failure"} 0.100328549
+snapshot_controller_operation_total_seconds_count{driver_name="driver4",operation_name="failure2",operation_status="Failure"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="0.1"} 0
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="0.25"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="0.5"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="1"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="2.5"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="5"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="10"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="15"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="30"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="60"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="120"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="300"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="600"} 1
+snapshot_controller_operation_total_seconds_bucket{driver_name="driver5",operation_name="unknown",operation_status="Unknown",le="+Inf"} 1
+snapshot_controller_operation_total_seconds_sum{driver_name="driver5",operation_name="unknown",operation_status="Unknown"} 0.100328549
+snapshot_controller_operation_total_seconds_count{driver_name="driver5",operation_name="unknown",operation_status="Unknown"} 1
+`
+	if err := verifyMetric(expected); err != nil {
+		t.Errorf("failed testing [%v]", err)
+	}
+}
+
+func verifyMetric(expected string) error {
```

Comment:
I think exposing the registry is a small price to pay for not having to duplicate ~100 lines of code that will have to be maintained over any changes in prometheus.

Reply:
IMO, exposing an unnecessary public interface for the sake of testing seems to be worse than a 100 lines more testing code :). I think its better to keep it hidden unless is necessary, and up till that point, we could switch to use the testutil.
Also, the test workflow is slightly different here, unit tests here are tested around http endpoint rather than prom registry.

Reply:
Sorry, as a likely maintainer for this code I'll have to insist. Duplicating complicated test verification is going to be a burden to maintain in the future. The public interface isn't a big deal, especially if this is going to be merged eventually with csi-lib-util.

I think the difference in the workflow doesn't prevent you from using the common library?

Reply:
chatted offline to resolve this, unlike csi-lib-util/metrics, unit tests can be deterministic as the latency will be passed in when recording a data point, the encapsulation of cache manages the latency calculation. It makes deterministic latency not easy in this case and thus not quite feasible to use GatherAndCompare.
however to @mattcary 's point, I have relaxed the latency checking from a fixed range to a greater than operation to avoid flake test resulted by latencies introduced from other component, i.e., prom lib.